File: 61-rotate-list/rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* flipHelper(ListNode* head){
        ListNode* prev = head;
        ListNode* cur = head->next;
        head->next = NULL;

        while (cur != NULL){
            ListNode* temp = cur->next;
            cur->next = prev;
            prev = cur;
            cur = temp;
        }

        return prev;
    }

    ListNode* shiftLeft(ListNode* head, ListNode* tail, int numShifts){
        while (numShifts){
            tail->next = head;
            tail = head;

            head = tail->next;
            tail->next = NULL;

            numShifts--;
        }

        return head;
    }

    ListNode* rotateRight(ListNode* head, int k) {
        if (head == NULL) return head;

        // Flip it, shift left, then flip back
        head = flipHelper(head);

        int numElements = 1;
        ListNode* tail = head;
        while (tail->next != NULL){
            tail = tail->next;
            numElements++;
        }
        int numShifts = k % numElements;

        head = shiftLeft(head, tail, numShifts);
        head = flipHelper(head);
    
        return head;
    }
};

File: 61-rotate-list/rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    # Pretty simple, just compute the new head and move previous chain to tail
    # Writing this out, its actually a bit annoying since there are a few steps

    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head:
            return None

        tail = head
        length = 1
        while (tail.next):
            length += 1
            tail = tail.next
        
        numShifts  = k % length
        if not numShifts:
            return head
        
        newHeadInd = length - numShifts
        newTail = head
        for i in range(newHeadInd - 1):
            newTail = newTail.next
        
        newHead = newTail.next
        newTail.next = None
        tail.next = head
        return newHead